### PR TITLE
.NET SDK for DiskRP Swagger 2021-04-01

### PR DESCRIPTION
**Pull request by rytruong**
Swagger PR: https://github.com/Azure/azure-rest-api-specs/pull/15665

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [x] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Disk.cs [hunk 1]
```diff
@@ -132,6 +140,7 @@ public Disk(string location, CreationData creationData, string id = default(stri
             OsType = osType;
             HyperVGeneration = hyperVGeneration;
             PurchasePlan = purchasePlan;
+            SupportedCapabilities = supportedCapabilities;
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Disk.cs:143]: need to add customization to preserve original constructor signature

Comment (grizzlytheodore) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Disk.cs:143]: please check customization folder

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DiskRestorePoint.cs [hunk 2]
```diff
@@ -54,18 +57,32 @@ public DiskRestorePoint()
...
+        /// operation.</param>
+        public DiskRestorePoint(string id = default(string), string name = default(string), string type = default(string), System.DateTime? timeCreated = default(System.DateTime?), string sourceResourceId = default(string), OperatingSystemTypes? osType = default(OperatingSystemTypes?), string hyperVGeneration = default(string), PurchasePlan purchasePlan = default(PurchasePlan), SupportedCapabilities supportedCapabilities = default(SupportedCapabilities), string familyId = default(string), string sourceUniqueId = default(string), Encryption encryption = default(Encryption), bool? supportsHibernation = default(bool?), string networkAccessPolicy = default(string), string publicNetworkAccess = default(string), string diskAccessId = default(string), double? completionPercent = default(double?))
             : base(id, name, type)
         {
             TimeCreated = timeCreated;
             SourceResourceId = sourceResourceId;
             OsType = osType;
             HyperVGeneration = hyperVGeneration;
             PurchasePlan = purchasePlan;
+            SupportedCapabilities = supportedCapabilities;
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DiskRestorePoint.cs:77]: customization needed

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DiskUpdate.cs [hunk 3]
```diff
@@ -99,8 +103,10 @@ public DiskUpdate(OperatingSystemTypes? osType = default(OperatingSystemTypes?),
             Tier = tier;
             BurstingEnabled = burstingEnabled;
             PurchasePlan = purchasePlan;
+            SupportedCapabilities = supportedCapabilities;
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DiskUpdate.cs:106]: customization needed

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Snapshot.cs [hunk 4]
```diff
@@ -96,6 +105,7 @@ public Snapshot(string location, CreationData creationData, string id = default(
             OsType = osType;
             HyperVGeneration = hyperVGeneration;
             PurchasePlan = purchasePlan;
+            SupportedCapabilities = supportedCapabilities;
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Snapshot.cs:108]: customization

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/SnapshotUpdate.cs [hunk 5]
```diff
@@ -64,6 +66,7 @@ public SnapshotUpdate(OperatingSystemTypes? osType = default(OperatingSystemType
             NetworkAccessPolicy = networkAccessPolicy;
             DiskAccessId = diskAccessId;
             SupportsHibernation = supportsHibernation;
+            PublicNetworkAccess = publicNetworkAccess;
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/SnapshotUpdate.cs:69]: customization

**Comment by grizzlytheodore** [comment]
two tests are failing in playback mode during CI check https://dev.azure.com/azure-sdk/public/_build/results?buildId=1094508&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774&t=3374e909-b3a0-5fde-4fba-1f1bb546637f

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Disk.cs (hunk 1 above)
Comment (rytruong) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Disk.cs:143]: added customization

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DiskRestorePoint.cs (hunk 2 above)
Comment (rytruong) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DiskRestorePoint.cs:77]: added customization

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DiskUpdate.cs (hunk 3 above)
Comment (rytruong) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/DiskUpdate.cs:106]: added customization

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Snapshot.cs (hunk 4 above)
Comment (rytruong) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/Snapshot.cs:108]: added customization

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/SnapshotUpdate.cs (hunk 5 above)
Comment (rytruong) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/SnapshotUpdate.cs:69]: added customization

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/Disk.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            ShareInfo = shareInfo;
+            NetworkAccessPolicy = networkAccessPolicy;
+            DiskAccessId = diskAccessId;
+            Tier = tier;
+            BurstingEnabled = burstingEnabled;
+            PropertyUpdatesInProgress = propertyUpdatesInProgress;
+            SupportsHibernation = supportsHibernation;
+            SecurityProfile = securityProfile;
+            CompletionPercent = completionPercent;
+            PublicNetworkAccess = publicNetworkAccess;
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/Disk.cs:143]: these newly added parameters should be removed

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/Disk.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        /// <param name="securityProfile">Contains the security related
+        /// information for the resource.</param>
+        /// <param name="supportedCapabilities">List of supported capabilities
+        /// for the image from which the OS disk was created.</param>
+        /// <param name="completionPercent">Percentage complete for the
+        /// background copy when a resource is created via the CopyStart
+        /// operation.</param>
+        /// <param name="publicNetworkAccess">Possible values include:
+        /// 'Enabled', 'Disabled'</param>
+        public Disk(string location, CreationData creationData, string id, string name, string type, IDictionary<string, string> tags, string managedBy, IList<string> managedByExtended, DiskSku sku, IList<string> zones, ExtendedLocation extendedLocation, System.DateTime? timeCreated, OperatingSystemTypes? osType,
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/Disk.cs:107]: the order of the parameters list should reflect the order of the parameters list in the default constructor before the new parameters were added

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/DiskUpdate.cs [hunk 8]
```diff
@@ -118,5 +117,28 @@ public DiskUpdate(OperatingSystemTypes? osType, int? diskSizeGB, EncryptionSetti
             Sku = sku;
             CustomInit();
         }
+
+        public DiskUpdate(OperatingSystemTypes? osType, int? diskSizeGB, EncryptionSettingsCollection encryptionSettingsCollection, long? diskIOPSReadWrite, long? diskMBpsReadWrite, long? diskIOPSReadOnly, long? diskMBpsReadOnly, int? maxShares, Encryption encryption, string networkAccessPolicy, string diskAccessId, string tier, bool? burstingEnabled, PurchasePlan purchasePlan, IDictionary<string, string> tags, DiskSku sku, SupportedCapabilities supportedCapabilities, string publicNetworkAccess)
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/DiskUpdate.cs:121]: the param list order is also different from default here

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/DiskUpdate.cs [hunk 9]
```diff
@@ -118,5 +117,28 @@ public DiskUpdate(OperatingSystemTypes? osType, int? diskSizeGB, EncryptionSetti
...
+            MaxShares = maxShares;
+            Encryption = encryption;
+            NetworkAccessPolicy = networkAccessPolicy;
+            DiskAccessId = diskAccessId;
+            Tier = tier;
+            BurstingEnabled = burstingEnabled;
+            PurchasePlan = purchasePlan;
+            Tags = tags;
+            Sku = sku;
+            SupportedCapabilities = supportedCapabilities;
```
Comment (grizzlytheodore) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/DiskUpdate.cs:139]: these new params should be removed

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/DiskUpdate.cs (hunk 8 above)
Comment (grizzlytheodore) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/DiskUpdate.cs:121]: please check other customizations

**Comment by grizzlytheodore** [comment]
@allenjzhang  the CI failure says, "internal build error"  could you help us resolve this? what do we need to do?

**Comment by grizzlytheodore** [comment]
@allenjzhang @m-nash  please review/merge

**Review by archerzz — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**